includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1522,4 +1569,446 @@ public function add_twentyfourteen_search() {
+		$css_selector = preg_replace( '/\s+/', ' ', $css_selector );
+
+		$xpath             = '';
+		$direct_descendant = false;
+		$token             = strtok( $css_selector, ' ' );
+
+		while ( false !== $token ) {
+			$matches = [];
+
+			// Direct descendant.
+			if ( preg_match( '/^>$/', $token, $matches ) ) {
+				$direct_descendant = true;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Single class.
+			if ( preg_match( '/^\.(?<class>[a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}*[ @class and contains( concat( ' ', normalize-space( @class ), ' ' ), ' {$matches['class']} ' ) ]";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Element.
+			if ( preg_match( '/^(?<element>[^.][a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}{$matches['element']}";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			$token = strtok( ' ' );
+		}
+
+		return $xpath;
+	}
+
+	/**
+	 * Register an AMP action to an event on a given element.
+	 *
+	 * If the element already contains one or more events or actions, the method
+	 * will assemble them in a smart way.
+	 *
+	 * @param DOMElement $element Element to add an action to.
+	 * @param string     $event   Event to trigger the action on.
+	 * @param string     $action  Action to add.
+	 */
+	protected function add_amp_action( DOMElement $element, $event, $action ) {
+		if ( ! $element->hasAttribute( 'on' ) ) {
+			$element->setAttribute( 'on', "{$event}:{$action}" );
+			return;
+		}
+
+		$events = explode( ';', $element->getAttribute( 'on' ) );
+
+		$found = false;
+		foreach ( $events as $index => $event_action_string ) {
+			if ( $found ) {
+				continue;
+			}
+
+			list( $existing_event, $existing_actions ) = explode( ':', $event_action_string, 2 );
+
+			if ( $existing_event !== $event ) {
+				continue;
+			}
+
+			$existing_actions_array   = explode( ',', $existing_actions );
+			$existing_actions_array[] = $action;
+			$actions                  = implode( ',', $existing_actions_array );
+
+			$events[ $index ] = "{$event}:{$actions}";
+			$found            = true;
+		}
+
+		if ( ! $found ) {
+			$events[] = "{$event}:${action}";
+		}
+
+		$element->setAttribute( 'on', implode( ';', $events ) );
+	}
+
+	/**
+	 * Check whether a given element has a specific class.
+	 *
+	 * @param DOMElement $element Element to check the classes of.
+	 * @param string     $class Class to check for.
+	 * @return bool Whether the element has the requested class.
+	 */
+	protected function has_class( DOMElement $element, $class ) {
+		if ( ! $element->hasAttribute( 'class' ) ) {
+			return false;
+		}
+
+		$classes = $element->getAttribute( 'class' );
+
+		return (bool) preg_match( "/\b{$class}\b/", $classes );
```

Comment:
To make this more robust:

```suggestion
		return (bool) preg_match( "/(^|\s){$class}(\s|$)/", $classes );
```

The `\b` will match hyphenated class names erroneously.

See also: 

https://github.com/ampproject/amp-wp/blob/cd1f8ae8bb5fcc89458767dde721797fe1158324/includes/sanitizers/class-amp-core-theme-sanitizer.php#L576-L582